lightning/src/routing/router.rs

```diff
@@ -333,84 +671,298 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
+	if payment_paths.len() > 50 {
+		payment_paths.truncate(50);
+	}
+
+	// Draw multiple sufficient routes by randomly combining the selected paths.
+	let mut drawn_routes = Vec::new();
+	for i in 0..payment_paths.len() {
+		let mut cur_route = Vec::<PaymentPath>::new();
+		let mut aggregate_route_value_msat = 0;
+
+		// Step (5).
+		// TODO: real random shuffle
+		// Currently just starts with i_th and goes up to i-1_th in a looped way.
+		let cur_payment_paths = [&payment_paths[i..], &payment_paths[..i]].concat();
+
+		// Step (6).
+		for payment_path in cur_payment_paths {
+			cur_route.push(payment_path.clone());
+			aggregate_route_value_msat += payment_path.get_value_msat();
+			if aggregate_route_value_msat > final_value_msat {
+				// Last path likely overpaid. Substract it from the most expensive
+				// (in terms of proportional fee) path in this route and recompute fees.
+				// This might be not the most economically efficient way, but fewer paths
+				// also makes routing more reliable.
+				let mut overpaid_value_msat = aggregate_route_value_msat - final_value_msat;
+
+				// First, drop some expensive low-value paths entirely if possible.
+				// Sort by value so that we drop many really-low values first, since
+				// fewer paths is better: the payment is less likely to fail.
+				// TODO: this could also be optimized by also sorting by feerate_per_sat_routed,
+				// so that the sender pays less fees overall. And also htlc_minimum_msat.
+				cur_route.sort_by_key(|path| path.get_value_msat());
+				// We should make sure that at least 1 path left.
+				let mut paths_left = cur_route.len();
+				cur_route.retain(|path| {
+					if paths_left == 1 {
+						return true
+					}
+					let mut keep = true;
+					let path_value_msat = path.get_value_msat();
+					if path_value_msat <= overpaid_value_msat {
+						keep = false;
+						overpaid_value_msat -= path_value_msat;
+						paths_left -= 1;
+					}
+					keep
+				});
+
+				if overpaid_value_msat == 0 {
+					break;
+				}
+
+				assert!(cur_route.len() > 0);
+
+				// Step (7).
+				// Now, substract the overpaid value from the most-expensive path.
+				// TODO: this could also be optimized by also sorting by feerate_per_sat_routed,
+				// so that the sender pays less fees overall. And also htlc_minimum_msat.
+				cur_route.sort_by_key(|path| { path.hops.iter().map(|hop| hop.channel_fees.proportional_millionths as u64).sum::<u64>() });
+				for expensive_payment_path in &mut cur_route {
+					if let Some(expensive_path_new_value_msat) = expensive_payment_path.get_value_msat().checked_sub(overpaid_value_msat) {
+						expensive_payment_path.update_value_and_recompute_fees(expensive_path_new_value_msat);
+						break;
+					} else {
+						// We already dropped all the small channels above, meaning all the
+						// remaining channels are larger than remaining overpaid_value_msat.
+						// Thus, this can't be negative.
+						unreachable!();
```

Comment:
nit: I dont think we need to bother with `checked_sub` if the other case is unreachable.

Reply:
the loop also seems to be unnecessary here, because first iteration immediately breaks from it.